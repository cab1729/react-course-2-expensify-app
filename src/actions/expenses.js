import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const removeExpense = ({ id = undefined }) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
});

export const startRemoveExpense = ({ id = undefined }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses/${id}`).remove()
        .then(() => {
            dispatch(removeExpense({ id }));
        })
        .catch((e) => {
            console.log('Error removing expense', e);
        }); 
    };
};

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates)
        .then(() => {
            dispatch(editExpense(id, updates));
        })
        .catch((e) => {
            console.log('Error updating expense', e);
        });
    };
};

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const expenses = [];
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({ 
                    id: childSnapshot.key, 
                    ...childSnapshot.val() 
                })
            });
            dispatch(setExpenses(expenses));
        });
        
    };
};