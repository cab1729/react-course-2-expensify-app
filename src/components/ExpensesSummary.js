import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getExpensesTotal from '../selectors/expenses-total.js';
import SelectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return(
        <div>
            <h1>Viewing {expensesCount} {expenseWord} totaling {formattedExpensesTotal}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = SelectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);