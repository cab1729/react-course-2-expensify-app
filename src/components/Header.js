import React from 'react';
import ReactDOM from 'react-dom';

import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <p><NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink></p>
        <p><NavLink to="/create" activeClassName="is-active">Add Expense</NavLink></p>
        <p><NavLink to="/help" activeClassName="is-active">Help</NavLink></p>
    </header>
);

export default Header;
