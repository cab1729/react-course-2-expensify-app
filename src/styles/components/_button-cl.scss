.button-cl {
    background-color: $blue;
    font-size: $font-size-large;
    font-weight: 300;
    color: white;
    padding: $s-size;
    border: none;
    display: inline-block;
    text-decoration: none;
    line-height: 1;
    margin-bottom: $s-size;
}

.button-cl--link {
    background: none;
}

.button-cl--secondary {
    background: #888888;
}