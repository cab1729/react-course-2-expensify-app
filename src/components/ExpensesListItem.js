import React from 'react';
import { connect } from 'react-redux';
import { Link, Route, BrowserRouter as Router } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { removeExpense } from '../actions/expenses';
import EditExpensePage from './EditExpensePage';

export const ExpensesListItem = ({dispatch, id, description, amount, createdAt, note}) => {
    return (
        <Router>
        <div>
            <Route path="/edit/:id" component={EditExpensePage} />
            <h3><Link to={`/edit/${id}`} >{description}</Link></h3>
            <p>Amount: {numeral(amount / 100).format('$0,0.00')} 
               -
               Date: {moment(createdAt).format('MMMM Do, YYYY')}</p>
            <p>{note}</p>
            
        </div>
        </Router>
    );
};

export default connect()(ExpensesListItem);