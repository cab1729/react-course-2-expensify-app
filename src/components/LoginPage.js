import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startLoginFB } from '../actions/auth';

export const LoginPage = ({ startLogin, startLoginFB }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify App</h1>
            <p>It's time to get your expenses under control.</p>
            <button className="button-cl" onClick={startLoginFB}>Login with Facebook</button>
            <button className="button-cl" onClick={startLogin}>Login with Google</button>
        </div>
    </div> 
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
    startLoginFB: () => dispatch(startLoginFB())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);