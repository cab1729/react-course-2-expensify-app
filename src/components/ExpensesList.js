import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpensesList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
        {
            props.expenses.length === 0 ? (
                <div className="list-item list-item--message">
                    <span>No Expenses</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return (<ExpensesListItem { ...expense } />);
                })
            )
        }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    console.log('ExpensesList: ', state.expenses, state.filters);
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesList);