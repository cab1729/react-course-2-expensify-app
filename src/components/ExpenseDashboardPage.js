import React from 'react';
import ReactDOM from 'react-dom';
import ExpensesList from './ExpensesList';
import ExpensesListFilters from './ExpensesListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesListFilters />
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;