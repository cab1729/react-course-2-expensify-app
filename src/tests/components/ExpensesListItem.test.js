import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesListItem } from '../../components/ExpensesListItem';
import expenses from '../fixtures/expenses';

test('Should render expenses list item with expense', () => {
    const wrapper = shallow(<ExpensesListItem {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});