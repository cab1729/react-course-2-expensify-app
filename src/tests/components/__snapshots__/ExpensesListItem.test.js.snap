// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expenses list item with expense 1`] = `
<BrowserRouter>
  <div>
    <Route
      component={[Function]}
      path="/edit/:id"
    />
    <h3>
      <Link
        replace={false}
        to="/edit/1"
      >
        Weed
      </Link>
    </h3>
    <p>
      Amount: 
      195000
      - Date: 
      December 31st, 1969
    </p>
    <p />
  </div>
</BrowserRouter>
`;
